Modules/Devices/src/Devices.Infrastructure/Persistence/Repository/OAuthClientsRepository.cs

```diff
@@ -13,21 +14,38 @@ namespace Backbone.Modules.Devices.Infrastructure.Persistence.Repository;
 public class OAuthClientsRepository : IOAuthClientsRepository
 {
     private readonly OpenIddictApplicationManager<CustomOpenIddictEntityFrameworkCoreApplication> _applicationManager;
+    private List<CustomOpenIddictEntityFrameworkCoreApplication> _cachedApplications;
 
     public OAuthClientsRepository(OpenIddictApplicationManager<CustomOpenIddictEntityFrameworkCoreApplication> applicationManager)
     {
         _applicationManager = applicationManager;
+        _cachedApplications = new List<CustomOpenIddictEntityFrameworkCoreApplication>();
     }
 
     public async Task<IEnumerable<OAuthClient>> FindAll(CancellationToken cancellationToken)
     {
-        var clients = await _applicationManager.ListAsync(applications => applications.Select(c => new OAuthClient(c.ClientId, c.DisplayName, c.TierId)), cancellationToken).ToListAsync(cancellationToken);
-        return clients;
+        var applications = await _applicationManager.ListAsync(cancellationToken: cancellationToken).ToListAsync(cancellationToken);
+
+        _cachedApplications = applications;
+
+        return applications.Select(client => new OAuthClient(client.ClientId!, client.DisplayName!, client.DefaultTier)).ToList();
     }
 
-    public async Task<CustomOpenIddictEntityFrameworkCoreApplication> Find(string clientId, CancellationToken cancellationToken)
+    public async Task<OAuthClient> Find(string clientId, CancellationToken cancellationToken, bool track = false)
     {
-        return await _applicationManager.FindByClientIdAsync(clientId, cancellationToken) ?? throw new NotFoundException(nameof(OAuthClient));
+        var application = await _applicationManager.FindByClientIdAsync(clientId, cancellationToken) ?? throw new NotFoundException(nameof(OAuthClient));
```

Comment:
Don't throw here. That's the job of the Handler. Just return null. And I know that we (and I include myself here) did that wrong in other places.

Reply:
What about the Update, Delete and ChangeClientSecret methods? Should they also have their errors moved to their handlers?

Reply:
For now we can leave it as it is there. We would need some more refactoring in order to change it there.